### ISSUE TITLE
Unkown state for sensor

### ISSUE THREAD

**Issue by AMajland** [jnxxx/homeassistant-dabblerdk_powermeterreader#2]
I have the MEP board running since early afternoon and the integration in HA a bit later.

The MEB dashboard is updating nicely on the ESP 
![image](https://user-images.githubusercontent.com/97510471/198896573-45a98975-cf76-4d19-8807-d2ace60d9ddc.png)

but the [sensor.echelon_energy_consumption] stopped updating after a couple of hours (the rest of the sensors was still updating)

I.e Power
![image](https://user-images.githubusercontent.com/97510471/198896550-77067d21-7215-43a8-a8b9-d43e1a6e1151.png)

I restarted HA and then it is updating again.

![image](https://user-images.githubusercontent.com/97510471/198896510-d7e0d041-e24c-4828-ab27-b60db9d30f53.png)

**Comment by jnxxx** [comment]
I think I have seen this once or twice, and made a fix that just wasn't published.
I haven't seen it since August, so now it is published.
Please update and try again.

The issue is that I have seen some spikes. Energy increases >1000 kWh, just to fall back on the next reading.
The way the entity is defined this is quite unfortunate. The integration will simply skip that reading and wait for the next one. It was this logic that had a bug.

If you continue to see the issue please enable some more logging. Debug may be to much for a longer period, but try warn first:
```
logger: 
  default: info
  logs: 
    custom_components.dabblerdk_powermeterreader: warning
```

**Comment by AMajland** [comment]
I've updated - and will wait for the issue to arise again before enabling more debug logging.

**Comment by japlab** [comment]
I updated tonight and are excited to see if it help avoiding those annoying ghost data.
Have a look at my Grafana. As you see the scaling is messed up caused by the spikes.
I report back in a couple of days about the result.
Best regards Jan

![Skærmbillede 2022-11-10 kl  21 04 51](https://user-images.githubusercontent.com/9386193/201195070-51bd8513-763a-4011-91c1-fba8a7add460.png)

**Comment by japlab** [comment]
> I updated tonight and are excited to see if it help avoiding those annoying ghost data. Have a look at my Grafana. As you see the scaling is messed up caused by the spikes. I report back in a couple of days about the result. Best regards Jan
> 
> ![Skærmbillede 2022-11-10 kl 21 04 51](https://user-images.githubusercontent.com/9386193/201195070-51bd8513-763a-4011-91c1-fba8a7add460.png)

**Comment by japlab** [comment]
2 day after the update I still get ghost readings from the MEP module.
L2 says  495360 Watt and sametime L1 is 57600 Watt.
 
![Skærmbillede 2022-11-12 kl  11 21 47](https://user-images.githubusercontent.com/9386193/201470074-ccc3193f-2a71-4294-9f74-64ae214680a1.png)

**Comment by jnxxx** [comment]
The fix mentioned above only looks at the accumulated energy (`Fwd_Act_Wh`).
It does not attempt to correct actual power readings (`L1_Fwd_W`, `L2_Fwd_W`, `L3_Fwd_W`).

I have however seen the issue you mention.
On my L3 I occasionally see power consumption near 65000W (suspiciously close to 16-bit limit).
I first figured it was an error because 230V x 13A fuse, then I shouldn't see more than 2990W.
However, I noticed it always happens when the compressor in my refrigerator starts. This could briefly pull some current when starting, possibly also without flipping the fuse if brief enough.
65000W is a lot though, and I do not know enough about fuses to say if possible.
With a motor (coil) and capacitor there is also the issue about phase shifting. Possibly that could also influence the readings, but I know very little about AC I must say.

If you enable warning logging, as shown above, it will log the raw readings from the MEP module when L3_Fwd_W is above 10000 W. This way I have validated that the readings are actually coming from the MEP module.

I have not tried to understand the code in the MEP firmware.

I guess it would be possible to make a workaround that ignores such surges.

**Comment by goegsig** [comment]
I have the same isssue  that the Energy consumption is Unavailable.

**Comment by jnxxx** [comment]
**Energy consumption**
I am uncertain if there have been two issues regarding Energy consumption becoming unavailable.
I thought I fixed it in the autumn, but I occasionally also saw it right after a reboot.
I have been investigating further and found an explanation for it: It always happened after a restart, but only if it failed to read data from the MEP module during initialization. The reason seems to be that the integration uses `Utility_SN` as part of the entities unique id. And this `Utility_SN` is not available when we do not have the data available.
For now it is fixed by postponing sensor initialization until we have successfully read data from the MEP module.
Home Assistant will retry 30 seconds later.

**Power spikes**
The power spikes are actually coming from the MEP module (or meter). I have been in doubt if it was real, or an error in the meter or the MEP module.
I noticed that the power of the three phases when summarized did not match the total power, which it normally should.
So now I make a validation and if it doesn't add up, I discard the read data and keep the previous data.
It is not easy to reproduce, but I have caught one occasion of it happening and overcoming it without ruining graph scaling.

Hopefully this will resolve these two issues.
As I have done some code refactoring as well, so I will do some more testing before publishing.
